crates/chia-consensus/src/gen/build_compressed_block.rs

```diff
@@ -0,0 +1,417 @@
+/// block generator with as many spends as possible, that fit within the
+/// specified maximum block cost. The priority of spends is typically the
+/// fee-per-cost (higher is better). The cost of the generated block is computed
+/// incrementally, based on the (compressed) size in bytes, the execution cost
+/// and conditions cost of each spend. The compressed size is not trivially
+/// predicted. Spends are added to the generator, and compressed, one at a time
+/// until we reach the target cost limit.
+#[cfg_attr(feature = "py-bindings", pyclass)]
+pub struct BlockBuilder {
+    allocator: Allocator,
+    signature: Signature,
+    sentinel: NodePtr,
+
+    // the cost of the block we've built up so far, not including the byte-cost.
+    // That's seprated out into the byte_cost member.
+    block_cost: u64,
+
+    // the byte cost, so for, of what's in the Serializer
+    byte_cost: u64,
+
+    // the number of spend bundles we've failed to add. Once this grows too
+    // large, we give up
+    num_skipped: u32,
+
+    // the serializer for the generator CLVM
+    ser: Serializer,
+}
+
+fn result(num_skipped: u32) -> BuildBlockResult {
+    if num_skipped > MAX_SKIPPED_ITEMS {
+        BuildBlockResult::Done
+    } else {
+        BuildBlockResult::KeepGoing
+    }
+}
+
+impl BlockBuilder {
+    pub fn new() -> io::Result<Self> {
+        let mut a = Allocator::new();
+
+        // the sentinel just needs to be a unique NodePtr. Since atoms may be
+        // de-duplicated (for small integers), we create a pair.
+        let sentinel = a.new_pair(NodePtr::NIL, NodePtr::NIL)?;
+
+        // the generator we produce is just a quoted list. Nothing fancy.
+        // Its format is as follows:
+        // (q . ( ( ( parent-id puzzle-reveal amount solution ) ... ) ) )
+
+        // the list of spends is the first (and only) item in an outer list
+        let spend_list = a.new_pair(sentinel, a.nil())?;
+        let quoted_list = a.new_pair(a.one(), spend_list)?;
+
+        let mut ser = Serializer::new(Some(sentinel));
+        ser.add(&a, quoted_list)?;
+
+        Ok(Self {
+            allocator: a,
+            signature: Signature::default(),
+            sentinel,
+            // TODO: where does this cost overhead come from?
+            block_cost: 20,
+            byte_cost: 0,
+            num_skipped: 0,
+            ser,
+        })
+    }
+
+    /// add a spend bundle to the generator. The cost must be *only* the CLVM
+    /// execution cost + the cost of the conditions. It must not include the byte cost
+    /// of the bundle. The byte cost is unpredictible as the generator is being
+    /// compressed. The true byte cost is computed by this function.
+    /// returns true if this bundle could be added to the generator, false otherwise
+    pub fn add_spend_bundle(
+        &mut self,
+        bundle: &SpendBundle,
+        cost: u64,
+        constants: &ConsensusConstants,
```

Comment:
Because these are constants, I would pass them to the class and simplify all the methods where they're needed.

Reply:
I like this idea, however, that would require a lifetime parameter on `BlockBuilder`, which isn't supported by pyo3 (kind of understandably).

Reply:
Hmm I see.